Sources/SparseSetModule/SparseSet+SparseStorage.swift

```diff
@@ -0,0 +1,184 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension SparseSet {
+  @usableFromInline
+  internal struct SparseStorage {
+    @usableFromInline
+    var _buffer: Buffer
+
+    @inlinable
+    @inline(__always)
+    init(_ buffer: Buffer) {
+      self._buffer = buffer
+    }
+
+    @inlinable
+    internal init<C: Collection>(withCapacity capacity: Int, keys: C) where C.Element == Key {
+      self._buffer = Buffer.bufferWith(capacity: capacity, keys: keys)
+    }
+
+    @inlinable
+    internal init(withCapacity capacity: Int) {
+      self._buffer = Buffer.bufferWith(capacity: capacity, keys: EmptyCollection<Key>())
+    }
+
+    @inlinable
+    internal init<C: Collection>(keys: C) where C.Element == Key {
+      let universeSize: Int = keys.max().map { Int($0) + 1 } ?? 0
+      self._buffer = Buffer.bufferWith(capacity: universeSize, keys: keys)
+    }
+  }
+}
+
+extension SparseSet.SparseStorage {
+  @usableFromInline
+  internal struct Header {
+    @usableFromInline
+    internal var capacity: Int
+
+    @inlinable
+    internal init(capacity: Int) {
+      self.capacity = capacity
+    }
+  }
+}
+
+extension SparseSet.SparseStorage {
+  @usableFromInline
+  internal final class Buffer: ManagedBuffer<Header, Key> {
+    /// Create a buffer populated with the given key data.
+    ///
+    /// - Parameters:
+    ///   - capacity: The capacity of the new buffer.
+    ///   - keys: A collection of keys.
+    ///
+    /// - Returns: A new buffer.
+    @inlinable
+    internal static func bufferWith<C: Collection>(capacity: Int, keys: C) -> Buffer where C.Element == Key {
+      assert(capacity >= keys.max().map { Int($0) + 1 } ?? 0)
+      let newBuffer = Buffer.create(
+        minimumCapacity: capacity,
+        makingHeaderWith: { _ in
+          Header(capacity: capacity)
+        })
+      newBuffer.withUnsafeMutablePointerToElements { ptr in
```

Comment:
Yes, the unoccupied positions of the sparse storage are uninitialized for better performance. If we filled the buffer with zeroes then initialization, resizing and copying (COW) operations will be slower. This might not matter much in practice though if users aren't performing these operations frequently.

What do you mean by a package configuration parameter?

Reply:
> What do you mean by a package configuration parameter?

Oh, I meant a compile-time conditional that we could use in `#if` statements, like the existing `COLLECTIONS_INTERNAL_CHECKS` one.

Reply:
Just adding a note here in case I forget:

If we are going to be fully initializing the sparse storage then there's a minor optimization we can consider for lookup operations. Details [here](https://skypjack.github.io/2020-08-02-ecs-baf-part-9/#lookup)